OpenRA.Mods.Common/FindActors.cs

```diff
@@ -0,0 +1,99 @@
+		{
+			// This line intersection check is done by first just finding all actors within a square that starts at the source, and ends at the target.
+			// Then we iterate over this list, and find all actors for which their health radius is at least within lineWidth of the line.
+			// For actors without a health radius, we simply check their center point.
+			// The square in which we select all actors must be large enough to encompass the entire line's width.
+			var xDir = Math.Sign(lineEnd.X - lineStart.X);
+			var yDir = Math.Sign(lineEnd.Y - lineStart.Y);
+
+			var dir = new WVec(xDir, yDir, 0);
+			var overselect = dir * (1024 + lineWidth.Length + targetExtraSearchRadius.Length);
+			var finalTarget = lineEnd + overselect;
+			var finalSource = lineStart - overselect;
+
+			var actorsInSquare = world.ActorMap.ActorsInBox(finalTarget, finalSource);
+			var intersectedActors = new List<Actor>();
+
+			foreach (var currActor in actorsInSquare)
+			{
+				long actorWidth = 0;
+				var healthInfo = currActor.Info.TraitInfoOrDefault<HealthInfo>();
+				if (healthInfo != null)
+					actorWidth = healthInfo.Radius.Length;
+
+				var projection = MinimumPointLineProjection(lineStart, lineEnd, currActor.CenterPosition);
+				var distance = (currActor.CenterPosition - projection).HorizontalLength;
+				var maxReach = (long)(actorWidth + lineWidth.Length);
+
+				if (distance <= maxReach)
+					intersectedActors.Add(currActor);
+			}
+
+			return intersectedActors;
+		}
+
+		/// <summary>
+		/// Find the point (D) on a line (A-B) that is closest to the target point (C).
+		/// </summary>
+		/// <param name="lineStart">The source point (tail) of the line</param>
+		/// <param name="lineEnd">The target point (head) of the line</param>
+		/// <param name="point">The target point that the minimum distance should be found to</param>
+		/// <returns>The WPos that is the point on the line that is closest to the target point</returns>
+		public static WPos MinimumPointLineProjection(WPos lineStart, WPos lineEnd, WPos point)
+		{
+			var squaredLength = (lineEnd - lineStart).HorizontalLengthSquared;
+
+			// Line has zero length, so just use the lineEnd position as the closest position.
+			if (squaredLength == 0)
+				return lineEnd;
+
+			// Consider the line extending the segment, parameterized as target + t (source - target).
+			// We find projection of point onto the line.
+			// It falls where t = [(point - target) . (source - target)] / |source - target|^2
+			// The normal DotProduct math would be (xDiff + yDiff) / dist, where dist = (target - source).LengthSquared;
+			// But in order to avoid floating points, we do not divide here, but rather work with the large numbers as far as possible.
+			// We then later divide by dist, only AFTER we have multiplied by the dotproduct.
+			var xDiff = ((long)point.X - (long)lineEnd.X) * ((long)lineStart.X - (long)lineEnd.X);
```

Comment:
You don't need to cast all of these.  One is enough (the rest will be promoted automatically).